prow/plugins/help/help.go

```diff
@@ -0,0 +1,150 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package help
+
+import (
+	"regexp"
+
+	"github.com/Sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "help"
+
+var (
+	helpLabel = "help-wanted"
+	helpRe    = regexp.MustCompile(`(?mi)^/help\s*$`)
+)
+
+type assignEvent struct {
+	action    string
+	body      string
+	login     string
+	org       string
+	repo      string
+	url       string
+	number    int
+	issue     github.Issue
+	assignees []github.User
+	hasLabel  func(label string) (bool, error)
+}
+
+func init() {
+	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
+	plugins.RegisterIssueHandler(pluginName, handleIssue)
+}
+
+type githubClient interface {
+	CreateComment(owner, repo string, number int, comment string) error
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+}
+
+func handleIssueComment(pc plugins.PluginClient, ic github.IssueCommentEvent) error {
+	// Only consider open issues.
+	if ic.Issue.IsPullRequest() || ic.Issue.State != "open" {
+		return nil
+	}
+
+	ae := assignEvent{
+		action:    ic.Action,
+		body:      ic.Comment.Body,
+		login:     ic.Comment.User.Login,
+		org:       ic.Repo.Owner.Login,
+		repo:      ic.Repo.Name,
+		url:       ic.Comment.HTMLURL,
+		number:    ic.Issue.Number,
+		issue:     ic.Issue,
+		assignees: ic.Issue.Assignees,
+		hasLabel:  func(label string) (bool, error) { return ic.Issue.HasLabel(label), nil },
+	}
+	return handle(pc.GitHubClient, pc.Logger, ae)
+}
+
+func handleIssue(pc plugins.PluginClient, i github.IssueEvent) error {
+	// Only consider open issues.
+	if i.Issue.IsPullRequest() || i.Issue.State != "open" {
+		return nil
+	}
+
+	ae := assignEvent{
+		action:    i.Action,
+		body:      i.Issue.Body,
+		login:     i.Issue.User.Login,
+		org:       i.Repo.Owner.Login,
+		repo:      i.Repo.Name,
+		url:       i.Issue.HTMLURL,
+		number:    i.Issue.Number,
+		issue:     i.Issue,
+		assignees: i.Issue.Assignees,
+		hasLabel:  func(label string) (bool, error) { return i.Issue.HasLabel(label), nil },
+	}
+	return handle(pc.GitHubClient, pc.Logger, ae)
+}
+
+func handle(gc githubClient, log *logrus.Entry, ae assignEvent) error {
+	// Determine if the issue has the help label
+	hasHelp, err := ae.hasLabel(helpLabel)
+	if err != nil {
+		log.WithError(err).Errorf("Unable to determine if request has %s label.", helpLabel)
+	}
+
+	// Determine if the issue has been assigned
+	isAssigned := false
+	if len(ae.assignees) != 0 {
```

Comment:
So the intent here is that this will react to the incoming github webhook event of an assign action.. no matter if that action is via `/assign` or if it's directly via github.

So if you assign something via `/assign`, what my intent would be is:
- `/help` command adds label to unassigned issue.
- User "a" `/assign`s it to themselves
- The assign prow plug in does it's thing and assigns the issue via the API
- Github sends a webhook event that User "a" has been assigned
- This triggers the help plugin that sends an API call to remove the label

Granted, this is intent, so it's possible that prow or Github isn't capable of this.